src/main/config.ts

```diff
@@ -0,0 +1,142 @@
+
+export type ConfigType = String | Boolean | Number;
+
+export interface ConfigDecl {
+    key: string;
+    type: ConfigType;
+    defaultValue?: string;
+}
+
+export interface ConfigOptions {
+    default?: ConfigType;
+}
+
+export type Configurable = { config: Config; }
+
+export function config(options: ConfigOptions = {}) {
+    return (prototype: Configurable, key: string) => {
+        const type = Reflect.getMetadata('design:type', prototype, key);
+        if (![String, Boolean, Number].includes(type)) {
+            throw new ConfigError('@config can only be used with string, number or boolean types');
+        }
+        const defaultValue = options.default == null ? undefined : String(options.default);
+        addClassMetadata<ConfigDecl>(CONFIG_METADATA_KEY, prototype, { key, type, defaultValue });
+        switch (type) {
+            case String: {
+                Object.defineProperty(prototype, key, {
+                    get() {
+                        return (this as Configurable).config.getString(key, defaultValue);
+                    }
+                });
+                break;
+            }
+            case Number: {
+                Object.defineProperty(prototype, key, {
+                    get() {
+                        return (this as Configurable).config.getNumber(key, defaultValue);
+                    }
+                });
+                break;
+            }
+            case Boolean: {
+                Object.defineProperty(prototype, key, {
+                    get() {
+                        return (this as Configurable).config.getBoolean(key, defaultValue);
+                    }
+                });
+                break;
+            }
+        }
+    };
+}
+
+/**
+ * Returns all `@config()` declared in class and its ancestors.
+ */
+export function getClassConfigs(classOrProto: any): ConfigDecl[] {
+    const target = classOrProto instanceof Function ? classOrProto.prototype : classOrProto;
+    return getClassMetadata(CONFIG_METADATA_KEY, target);
+}
+
+/**
+ * Returns all `@config()` declarations for all classes bound to the container.
+ */
+export function getContainerConfigs(container: Container): ConfigDecl[] {
+    const result: ConfigDecl[] = [];
+    for (const bindings of getBindingsMap(container).values()) {
+        for (const binding of bindings) {
+            if (typeof binding.implementationType === 'function') {
+                const configs = getClassConfigs(binding.implementationType);
+                result.push(...configs);
+            }
+        }
+    }
+    return result.sort((a, b) => a.key > b.key ? 1 : -1);
+}
+
+@injectable()
+export abstract class Config {
+    abstract resolve(key: string): string | null;
+
+    protected getOrNull<T>(
+        key: string,
+        parse: (str: string) => T | null,
+        defaultValue?: string | T,
+    ): T | null {
+        let str = this.resolve(key);
+        if (str == null) {
+            str = String(defaultValue);
```

Comment:
if defaultValue is `undefined`, would it parse `'undefined'` if the `parse` argument is `parseString`?

Reply:
Fixed this one, thanks 🦅 👁️